File: dashboards/dashboards/page/manufacture_dashboard/manufacture_dashboard.js
frappe.pages["manufacture-dashboard"].on_page_load = function (wrapper) {
	new dashboards.ui.ManufactureDashboardPage(wrapper);
};

frappe.provide("dashboards.ui");

dashboards.ui.ManufactureDashboardPage = class ManufactureDashboardPage {
	constructor(wrapper) {
		this.wrapper = $(wrapper);
		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: __("Производственный дашборд"),
			single_column: true,
		});

		this.months = [
			{ key: "Jan", label: "Янв", full: "Январь" },
			{ key: "Feb", label: "Фев", full: "Февраль" },
			{ key: "Mar", label: "Мар", full: "Март" },
			{ key: "Apr", label: "Апр", full: "Апрель" },
			{ key: "May", label: "Май", full: "Май" },
			{ key: "Jun", label: "Июн", full: "Июнь" },
			{ key: "Jul", label: "Июл", full: "Июль" },
			{ key: "Aug", label: "Авг", full: "Август" },
			{ key: "Sep", label: "Сен", full: "Сентябрь" },
			{ key: "Oct", label: "Окт", full: "Октябрь" },
			{ key: "Nov", label: "Ноя", full: "Ноябрь" },
			{ key: "Dec", label: "Дек", full: "Декабрь" },
		];
		this.years = [];
		this.state = {
			year: "",
			month: "",
		};
		this.data = null;

		this.make_layout();
		this.bind_events();
		this.load_data();
	}

	make_layout() {
		this.wrapper.find(".layout-main-section-wrapper").addClass("manufacture-dashboard-layout");
		this.wrapper.find(".page-head").addClass("manufacture-dashboard-page-head");
		this.page.main.removeClass("frappe-card");

		this.page.main.html(`
			<div class="mfd-screen">
				<section class="mfd-main">
					<div class="mfd-filters">
						<div class="mfd-filter-label">${__("ФИЛЬТР ГОДА")}</div>
						<div class="mfd-year-select">
							<button class="mfd-select" type="button" data-year-toggle aria-expanded="false">
								<span data-region="selected-year"></span>
								<span class="mfd-chevron" aria-hidden="true"></span>
							</button>
							<div class="mfd-year-menu" data-region="year-menu"></div>
						</div>
						<div class="mfd-filter-label">${__("ФИЛЬТР МЕСЯЦА")}</div>
						<div class="mfd-month-grid" data-region="month-grid"></div>
					</div>

					<div class="mfd-kpi-grid">
						<section class="mfd-card mfd-kpi-card mfd-kpi-card--production">
							<div class="mfd-card-title">
								<span class="mfd-title-icon mfd-title-icon--blue" aria-hidden="true">▧</span>
								<h2>${__("Производство за текущий месяц")}</h2>
							</div>
							<div class="mfd-kpi-row">
								<strong data-region="month-production"></strong>
							</div>
						</section>
						<section class="mfd-card mfd-kpi-card mfd-kpi-card--extra-expense">
							<div class="mfd-card-title">
								<span class="mfd-title-icon mfd-title-icon--green" aria-hidden="true">⌑</span>
								<h2>${__("Доп. расход")}</h2>
							</div>
							<div class="mfd-kpi-row">
								<strong data-region="dop-rasxod-total"></strong>
							</div>
						</section>
						<section class="mfd-card mfd-kpi-card mfd-kpi-card--cost">
							<div class="mfd-card-title">
								<span class="mfd-title-icon mfd-title-icon--violet" aria-hidden="true">⌁</span>
								<h2>${__("Сумма себестоимости")}</h2>
							</div>
							<div class="mfd-kpi-row">
								<strong data-region="average-cost"></strong>
							</div>
						</section>
					</div>

					<div class="mfd-chart-grid">
						<section class="mfd-card mfd-volume-card">
							<div class="mfd-card-head">
								<div class="mfd-card-title">
									<span class="mfd-title-icon mfd-title-icon--blue" aria-hidden="true">⌞</span>
									<h2>${__("Годовой объем производства")}</h2>
								</div>
								<div class="mfd-legend">
									<span><i class="is-blue"></i>${__("Тонны")}</span>
								</div>
							</div>
							<div class="mfd-combo-chart" data-region="combo-chart"></div>
						</section>

						<section class="mfd-card mfd-production-table-card">
							<div class="mfd-card-title">
								<span class="mfd-title-icon mfd-title-icon--green" aria-hidden="true">☷</span>
								<h2>${__("Произведенные товары за месяц")}</h2>
							</div>
							<div class="mfd-production-table-wrap" data-region="production-table"></div>
						</section>
					</div>
					<div class="mfd-updated" data-region="updated-at"></div>
				</section>
			</div>
		`);

		dashboards.ui.setupDashboardSidebar({
			page: this.page,
			route: "manufacture-dashboard",
		});

		this.$selectedYear = this.page.main.find('[data-region="selected-year"]');
		this.$yearMenu = this.page.main.find('[data-region="year-menu"]');
		this.$monthGrid = this.page.main.find('[data-region="month-grid"]');
		this.$monthProduction = this.page.main.find('[data-region="month-production"]');
		this.$dopRasxodTotal = this.page.main.find('[data-region="dop-rasxod-total"]');
		this.$averageCost = this.page.main.find('[data-region="average-cost"]');
		this.$comboChart = this.page.main.find('[data-region="combo-chart"]');
		this.$productionTable = this.page.main.find('[data-region="production-table"]');
		this.$updatedAt = this.page.main.find('[data-region="updated-at"]');
	}

	bind_events() {
		this.page.main.off(".manufacture-dashboard");

		this.page.main.on("click.manufacture-dashboard", "[data-year-toggle]", (event) => {
			event.stopPropagation();
			const $select = $(event.currentTarget).closest(".mfd-year-select");
			const isOpen = $select.hasClass("is-open");
			$select.toggleClass("is-open", !isOpen);
			$(event.currentTarget).attr("aria-expanded", isOpen ? "false" : "true");
		});

		this.page.main.on("click.manufacture-dashboard", "[data-year]", (event) => {
			const year = String($(event.currentTarget).data("year"));
			this.page.main.find(".mfd-year-select").removeClass("is-open");
			this.page.main.find("[data-year-toggle]").attr("aria-expanded", "false");
			this.load_data({ year, month: this.state.month });
		});

		this.page.main.on("click.manufacture-dashboard", "[data-month]", (event) => {
			const month = String($(event.currentTarget).data("month"));
			this.load_data({ year: this.state.year, month });
		});

		$(document).off("click.manufacture-dashboard-year").on("click.manufacture-dashboard-year", (event) => {
			if ($(event.target).closest(".mfd-year-select").length) {
				return;
			}

			this.page.main.find(".mfd-year-select").removeClass("is-open");
			this.page.main.find("[data-year-toggle]").attr("aria-expanded", "false");
		});
	}

	load_data(filters = {}) {
		this.show_loading();
		return frappe
			.call({
				method: "dashboards.dashboards.page.manufacture_dashboard.manufacture_dashboard.get_dashboard_data",
				args: {
					year: filters.year || this.state.year || undefined,
					month: filters.month || this.state.month || undefined,
				},
			})
			.then((response) => {
				this.data = response.message || {};
				const backendFilters = this.data.filters || {};
				this.years = backendFilters.years || [];
				this.state.year = backendFilters.selected_year || "";
				this.state.month = backendFilters.selected_month || "";
				this.render();
			})
			.catch(() => {
				frappe.msgprint(__("Не удалось загрузить данные производственного дашборда."));
				this.render_empty();
			});
	}

	show_loading() {
		this.$selectedYear.text(this.state.year || "...");
		this.$yearMenu.empty();
		this.$monthGrid.html(this.months.map((month) => this.render_month_button(month)).join(""));
		const loadingMarkup = `<div class="mfd-loading">${__("Загрузка...")}</div>`;
		this.$monthProduction.text("...");
		this.$dopRasxodTotal.text("...");
		this.$averageCost.text("...");
		this.$comboChart.html(loadingMarkup);
		this.$productionTable.html(loadingMarkup);
	}

	render_empty() {
		this.$comboChart.html(`<div class="mfd-loading">${__("Нет данных")}</div>`);
		this.$productionTable.html(`<div class="mfd-loading">${__("Нет данных")}</div>`);
	}

	render() {
		const summary = this.data?.summary || {};
		this.$selectedYear.text(this.state.year);
		this.$yearMenu.html(this.years.map((year) => this.render_year_option(year)).join(""));
		this.$monthGrid.html(this.months.map((month) => this.render_month_button(month)).join(""));
		this.$monthProduction.text(summary.monthProductionDisplay || `0 ${__("T")}`);
		this.$dopRasxodTotal.text(summary.dopRasxodTotalDisplay || "0 UZS");
		this.$averageCost.text(summary.summaSebTotalDisplay || "0 UZS");
		this.$updatedAt.text(`${__("Последнее обновление")}: ${this.data?.updated_at || "—"}`);
		this.render_combo_chart();
		this.render_production_table();
	}

	render_year_option(year) {
		return `<button class="mfd-year-option ${year === this.state.year ? "is-active" : ""}" type="button" data-year="${year}">${frappe.utils.escape_html(year)}</button>`;
	}

	render_month_button(month) {
		return `<button class="mfd-month ${month.key === this.state.month ? "is-active" : ""}" type="button" data-month="${month.key}">${frappe.utils.escape_html(month.label)}</button>`;
	}

	render_combo_chart() {
		const rows = this.data?.monthly || [];
		if (!rows.length) {
			this.$comboChart.html(`<div class="mfd-loading">${__("Нет данных")}</div>`);
			return;
		}
		const maxTon = Math.max(1, ...rows.map((row) => Number(row.ton || 0)));

		this.$comboChart.html(`
			<div class="mfd-y-axis mfd-y-axis--left">
				${this.build_axis_labels(maxTon, 8, (value) => this.format_number(value)).map((label) => `<span>${label}</span>`).join("")}
			</div>
			<div class="mfd-bars">
				${rows
					.map((row) => {
						const month = this.months.find((item) => item.key === row.month);
						return `
							<div class="mfd-bar-item">
								<span class="mfd-bar-value">${frappe.utils.escape_html(row.ton_display || `${this.format_number(row.ton)} T`)}</span>
								<div class="mfd-bar" style="height: ${Math.max(2, Math.round((Number(row.ton || 0) / maxTon) * 92))}%"></div>
								<span class="mfd-bar-label">${frappe.utils.escape_html(month ? month.full : row.month)}</span>
							</div>
						`;
					})
					.join("")}
			</div>
		`);
	}

	render_production_table() {
		const rows = this.data?.production_rows || [];
		if (!rows.length) {
			this.$productionTable.html(`<div class="mfd-loading">${__("Нет данных")}</div>`);
			return;
		}

		const total = rows.reduce(
			(accumulator, row) => {
				accumulator.ton += Number(row.ton_raw ?? row.ton ?? 0);
				accumulator.summaSeb += Number(row.summa_seb_raw ?? row.summa_seb ?? 0);
				accumulator.dopRasxod += Number(row.dop_rasxod_raw ?? row.dop_rasxod ?? 0);
				return accumulator;
			},
			{ ton: 0, summaSeb: 0, dopRasxod: 0 }
		);

		this.$productionTable.html(`
			<table class="mfd-production-table">
				<thead>
					<tr>
						<th class="is-text">${__("Наименование товара")}</th>
						<th class="is-number">${__("Тонны")}</th>
						<th class="is-number">${__("Сумма себестоимости")}</th>
						<th class="is-number">${__("Доп. расход")}</th>
					</tr>
				</thead>
				<tbody>
					${rows
						.map(
							(row) => `
								<tr>
									<td class="is-text">${frappe.utils.escape_html(row.label || "")}</td>
									<td class="is-number">${frappe.utils.escape_html(row.ton_display || `${this.format_decimal(row.ton, 2)} T`)}</td>
									<td class="is-number">${frappe.utils.escape_html(row.summa_seb_display || this.format_number(row.summa_seb))}</td>
									<td class="is-number">${frappe.utils.escape_html(row.dop_rasxod_display || this.format_number(row.dop_rasxod))}</td>
								</tr>
							`
						)
						.join("")}
				</tbody>
				<tfoot>
					<tr>
						<td class="is-text">${__("Итого")}</td>
						<td class="is-number">${frappe.utils.escape_html(`${this.format_decimal(total.ton, 2)} T`)}</td>
						<td class="is-number">${frappe.utils.escape_html(`${this.format_number(total.summaSeb)} UZS`)}</td>
						<td class="is-number">${frappe.utils.escape_html(`${this.format_number(total.dopRasxod)} UZS`)}</td>
					</tr>
				</tfoot>
			</table>
		`);
	}

	build_axis_labels(maxValue, count, formatter) {
		const max = Math.max(Number(maxValue || 0), 1);
		return Array.from({ length: count }, (_, index) => {
			const value = max - (max / (count - 1)) * index;
			return formatter(value);
		});
	}

	format_compact_money(value) {
		const amount = Number(value || 0);
		const absolute = Math.abs(amount);
		if (absolute >= 1_000_000_000) {
			return `${this.format_decimal(amount / 1_000_000_000, 1)}B`;
		}
		if (absolute >= 1_000_000) {
			return `${this.format_decimal(amount / 1_000_000, 1)}M`;
		}
		if (absolute >= 1_000) {
			return `${this.format_decimal(amount / 1_000, 1)}K`;
		}
		return this.format_number(amount);
	}

	format_decimal(value, maximumFractionDigits = 1) {
		return new Intl.NumberFormat("en-US", {
			maximumFractionDigits,
			minimumFractionDigits: 0,
		}).format(value || 0);
	}

	format_number(value) {
		return new Intl.NumberFormat("en-US", { maximumFractionDigits: 0 }).format(value || 0);
	}
};

File: dashboards/dashboards/page/manufacture_dashboard/manufacture_dashboard.py
from __future__ import annotations

from functools import lru_cache
from typing import Any

import frappe
from frappe.utils import cint, flt, getdate, now_datetime, today

from dashboards.dashboards.dashboard_data import (
    REPORTING_CURRENCY,
    convert_company_currency_amount,
    format_number,
    get_default_company,
)
from dashboards.dashboards.page.main_dashboard.main_dashboard import (
    _safe_div,
    _to_tons,
)


SHORT_MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]
MONTH_INDEX = {label: index + 1 for index, label in enumerate(SHORT_MONTHS)}


def get_context(context):
    context.no_cache = 1


def _month_no(month: str | int | None) -> int | None:
    if month in (None, ""):
        return None
    if isinstance(month, int):
        return month if 1 <= month <= 12 else None

    value = str(month).strip()
    if value.isdigit():
        parsed = cint(value)
        return parsed if 1 <= parsed <= 12 else None

    return MONTH_INDEX.get(value[:3].title())


def _format_uzs(value: float, precision: int = 0) -> str:
    return f"{format_number(value, precision=precision)} UZS"


def _format_tons(value: float) -> str:
    return f"{format_number(value, precision=2)} T"


def _empty_month_map() -> dict[int, dict[str, float]]:
    return {
        month_no: {
            "manufactured_qty": 0.0,
            "manufactured_tons": 0.0,
            "manufacturing_cost": 0.0,
            "raw_cost": 0.0,
            "raw_qty": 0.0,
            "cost_per_ton": 0.0,
            "raw_cost_per_qty": 0.0,
        }
        for month_no in range(1, 13)
    }


def _manufacture_condition(alias: str = "se") -> str:
    return f"({alias}.stock_entry_type = 'Manufacture' OR {alias}.purpose = 'Manufacture')"


@lru_cache(maxsize=12)
def _get_monthly_indirect_expense_from_profit_and_loss(year: str) -> dict[int, float]:
    company = get_default_company()
    if not company:
        return {month_no: 0.0 for month_no in range(1, 13)}

    latest_posting_date = frappe.db.sql(
        """
        SELECT MAX(posting_date) AS posting_date
        FROM `tabGL Entry`
        WHERE company = %(company)s
          AND YEAR(posting_date) = %(year)s
          AND docstatus = 1
          AND is_cancelled = 0
        """,
        {"company": company, "year": cint(year)},
        as_dict=True,
    )[0].posting_date

    if not latest_posting_date:
        return {month_no: 0.0 for month_no in range(1, 13)}

    from erpnext.accounts.report.profit_and_loss_statement import profit_and_loss_statement
    from erpnext.accounts.utils import get_fiscal_year

    fiscal_year = get_fiscal_year(latest_posting_date, company=company)[0]
    filters = frappe._dict(
        {
            "company": company,
            "from_fiscal_year": fiscal_year,
            "to_fiscal_year": fiscal_year,
            "period_start_date": f"{cint(year)}-01-01",
            "period_end_date": str(getdate(latest_posting_date)),
            "filter_based_on": "Date Range",
            "periodicity": "Monthly",
            "accumulated_values": 0,
            "include_default_book_entries": 1,
            "presentation_currency": REPORTING_CURRENCY,
        }
    )
    columns, data, *_rest = profit_and_loss_statement.execute(filters)

    target_rows = [
        row
        for row in data
        if "Indirect Expenses" in str(row.get("account") or row.get("account_name") or "")
        or "Indirect Expenses" in str(row.get("account_name") or "")
    ]
    target_row = sorted(target_rows, key=lambda row: flt(row.get("indent")))[0] if target_rows else None

    month_map = {month_no: 0.0 for month_no in range(1, 13)}
    if not target_row:
        return month_map

    month_lookup = {
        "jan": 1,
        "feb": 2,
        "mar": 3,
        "apr": 4,
        "may": 5,
        "jun": 6,
        "jul": 7,
        "aug": 8,
        "sep": 9,
        "oct": 10,
        "nov": 11,
        "dec": 12,
    }
    for column in columns:
        fieldname = str(column.get("fieldname") or "")
        if "_" not in fieldname:
            continue
        month_no = month_lookup.get(fieldname.split("_", 1)[0])
        if month_no:
            month_map[month_no] = flt(target_row.get(fieldname))

    return month_map


@lru_cache(maxsize=12)
def _get_dashboard_years() -> list[str]:
    rows = frappe.db.sql(
        f"""
        SELECT DISTINCT YEAR(se.posting_date) AS year
        FROM `tabStock Entry` se
        WHERE se.docstatus = 1
          AND {_manufacture_condition("se")}
          AND se.posting_date IS NOT NULL
        ORDER BY year
        """,
        as_dict=True,
    )
    years = [str(row.year) for row in rows if row.year]
    if years:
        return years
    return [str(getdate(today()).year)]


def _get_latest_posting_date(year: str) -> Any:
    row = frappe.db.sql(
        f"""
        SELECT MAX(se.posting_date) AS posting_date
        FROM `tabStock Entry` se
        WHERE se.docstatus = 1
          AND {_manufacture_condition("se")}
          AND YEAR(se.posting_date) = %(year)s
        """,
        {"year": cint(year)},
        as_dict=True,
    )[0]
    return row.posting_date


def _resolve_filters(year: str | None = None, month: str | None = None) -> dict[str, Any]:
    years = _get_dashboard_years()
    selected_year = str(year) if year and str(year) in years else years[-1]
    latest_posting_date = _get_latest_posting_date(selected_year)
    default_month = SHORT_MONTHS[getdate(latest_posting_date).month - 1] if latest_posting_date else SHORT_MONTHS[0]
    selected_month = month[:3].title() if month and month[:3].title() in MONTH_INDEX else default_month
    return {
        "years": years,
        "months": SHORT_MONTHS,
        "selected_year": selected_year,
        "selected_month": selected_month,
    }


@lru_cache(maxsize=12)
def _get_monthly_manufacture_metrics(year: str) -> dict[int, dict[str, float]]:
    rows = frappe.db.sql(
        f"""
        SELECT
            entry_month,
            posting_date,
            company,
            SUM(manufactured_qty) AS manufactured_qty,
            SUM(raw_cost) AS raw_cost,
            SUM(raw_qty) AS raw_qty,
            SUM(additional_cost) AS additional_cost
        FROM (
            SELECT
                se.name,
                MONTH(se.posting_date) AS entry_month,
                se.posting_date,
                se.company,
                SUM(CASE WHEN COALESCE(sed.is_finished_item, 0) = 1 THEN COALESCE(sed.qty, 0) ELSE 0 END) AS manufactured_qty,
                SUM(CASE WHEN COALESCE(sed.is_finished_item, 0) = 0 THEN ABS(COALESCE(sed.basic_amount, 0)) ELSE 0 END) AS raw_cost,
                SUM(CASE WHEN COALESCE(sed.is_finished_item, 0) = 0 THEN ABS(COALESCE(sed.qty, 0)) ELSE 0 END) AS raw_qty,
                MAX(COALESCE(se.total_additional_costs, 0)) AS additional_cost
            FROM `tabStock Entry` se
            LEFT JOIN `tabStock Entry Detail` sed ON sed.parent = se.name
            WHERE se.docstatus = 1
              AND {_manufacture_condition("se")}
              AND YEAR(se.posting_date) = %(year)s
            GROUP BY se.name, MONTH(se.posting_date), se.posting_date, se.company
        ) manufacture_entries
        GROUP BY entry_month, posting_date, company
        """,
        {"year": cint(year)},
        as_dict=True,
    )

    month_map = _empty_month_map()
    for row in rows:
        month_no = cint(row.entry_month)
        if month_no not in month_map:
            continue

        raw_cost = convert_company_currency_amount(row.raw_cost, row.posting_date, row.company)
        additional_cost = convert_company_currency_amount(row.additional_cost, row.posting_date, row.company)
        total_cost = raw_cost + additional_cost
        month_map[month_no]["manufactured_qty"] += flt(row.manufactured_qty)
        month_map[month_no]["manufacturing_cost"] += total_cost
        month_map[month_no]["raw_cost"] += raw_cost
        month_map[month_no]["raw_qty"] += flt(row.raw_qty)

    for month_no, values in month_map.items():
        values["manufactured_tons"] = _to_tons(values["manufactured_qty"])
        values["cost_per_ton"] = _safe_div(values["manufacturing_cost"], values["manufactured_tons"])
        values["raw_cost_per_qty"] = _safe_div(values["raw_cost"], _to_tons(values["raw_qty"]))

    return month_map


@lru_cache(maxsize=24)
def _get_production_table_rows(year: str, month: str | None = None) -> list[dict[str, Any]]:
    month_no = _month_no(month)
    month_clause = "AND MONTH(se.posting_date) = %(month)s" if month_no else ""
    params = {"year": cint(year)}
    if month_no:
        params["month"] = month_no

    rows = frappe.db.sql(
        f"""
        SELECT
            COALESCE(NULLIF(sed.item_code, ''), NULLIF(sed.item_name, ''), 'Неизвестный товар') AS item_code,
            COALESCE(NULLIF(sed.item_name, ''), sed.item_code, 'Неизвестный товар') AS item_label,
            se.posting_date,
            se.company,
            SUM(COALESCE(sed.qty, 0)) AS manufactured_qty,
            SUM(
                CASE
                    WHEN COALESCE(sed.amount, 0) != 0 THEN COALESCE(sed.amount, 0)
                    WHEN COALESCE(sed.basic_amount, 0) != 0 THEN COALESCE(sed.basic_amount, 0)
                    ELSE COALESCE(sed.qty, 0) * COALESCE(sed.valuation_rate, 0)
                END
            ) AS manufacturing_amount
        FROM `tabStock Entry` se
        INNER JOIN `tabStock Entry Detail` sed ON sed.parent = se.name
        WHERE se.docstatus = 1
          AND {_manufacture_condition("se")}
          AND COALESCE(sed.is_finished_item, 0) = 1
          AND YEAR(se.posting_date) = %(year)s
          {month_clause}
        GROUP BY
            COALESCE(NULLIF(sed.item_code, ''), NULLIF(sed.item_name, ''), 'Неизвестный товар'),
            COALESCE(NULLIF(sed.item_name, ''), sed.item_code, 'Неизвестный товар'),
            se.posting_date,
            se.company
        """,
        params,
        as_dict=True,
    )

    grouped: dict[str, dict[str, Any]] = {}
    for row in rows:
        item_key = str(row.item_code or row.item_label or "Неизвестный товар")
        if item_key not in grouped:
            grouped[item_key] = {
                "label": str(row.item_label or item_key),
                "qty": 0.0,
                "summa_seb": 0.0,
            }

        grouped[item_key]["qty"] += flt(row.manufactured_qty)
        grouped[item_key]["summa_seb"] += convert_company_currency_amount(
            row.manufacturing_amount,
            row.posting_date,
            row.company,
        )

    result = []
    for values in grouped.values():
        tons = _to_tons(values["qty"])
        summa_seb = flt(values["summa_seb"])
        result.append(
            {
                "label": values["label"],
                "ton": round(tons, 2),
                "ton_raw": tons,
                "ton_display": _format_tons(tons),
                "summa_seb": round(summa_seb, 2),
                "summa_seb_raw": summa_seb,
                "summa_seb_display": _format_uzs(summa_seb),
            }
        )

    return sorted(result, key=lambda row: (-flt(row["ton"]), row["label"]))


def _add_dop_rasxod_rows(rows: list[dict[str, Any]], indirect_total: float) -> list[dict[str, Any]]:
    month_total_tons = sum(flt(row.get("ton_raw", row.get("ton"))) for row in rows)
    for row in rows:
        dop_rasxod = _safe_div(flt(row.get("ton_raw", row.get("ton"))) * flt(indirect_total), month_total_tons)
        row["dop_rasxod"] = round(dop_rasxod, 2)
        row["dop_rasxod_raw"] = dop_rasxod
        row["dop_rasxod_display"] = _format_uzs(dop_rasxod)
    return rows



def _build_summary(month: str, metrics: dict[int, dict[str, float]]) -> dict[str, Any]:
    month_no = _month_no(month) or 1
    selected_month = metrics[month_no]

    return {
        "monthProduction": round(selected_month["manufactured_tons"], 2),
        "monthProductionDisplay": _format_tons(selected_month["manufactured_tons"]),
        "averageCost": round(selected_month["cost_per_ton"], 2),
        "averageCostDisplay": _format_uzs(selected_month["cost_per_ton"]),
    }


def _build_monthly_series(metrics: dict[int, dict[str, float]]) -> list[dict[str, Any]]:
    series = []
    for month_no, month in enumerate(SHORT_MONTHS, start=1):
        values = metrics[month_no]
        series.append(
            {
                "month": month,
                "ton": round(values["manufactured_tons"], 2),
                "ton_display": _format_tons(values["manufactured_tons"]),
                "cost": round(values["cost_per_ton"], 2),
                "cost_display": _format_uzs(values["cost_per_ton"]),
            }
        )
    return series



@frappe.whitelist()
def get_dashboard_data(year: str | None = None, month: str | None = None) -> dict[str, Any]:
    _get_dashboard_years.cache_clear()
    _get_monthly_indirect_expense_from_profit_and_loss.cache_clear()
    _get_monthly_manufacture_metrics.cache_clear()
    _get_production_table_rows.cache_clear()
    filters = _resolve_filters(year, month)
    selected_year = filters["selected_year"]
    selected_month = filters["selected_month"]
    metrics = _get_monthly_manufacture_metrics(selected_year)
    production_rows = _get_production_table_rows(selected_year, selected_month)
    indirect_total = flt(_get_monthly_indirect_expense_from_profit_and_loss(selected_year).get(_month_no(selected_month) or 0))
    production_rows = _add_dop_rasxod_rows(production_rows, indirect_total)
    summa_seb_total = sum(flt(row.get("summa_seb_raw", row.get("summa_seb"))) for row in production_rows)
    dop_rasxod_total = sum(flt(row.get("dop_rasxod_raw", row.get("dop_rasxod"))) for row in production_rows)
    summary = _build_summary(selected_month, metrics)
    summary.update(
        {
            "summaSebTotal": round(summa_seb_total, 2),
            "summaSebTotalDisplay": _format_uzs(summa_seb_total),
            "dopRasxodTotal": round(dop_rasxod_total, 2),
            "dopRasxodTotalDisplay": _format_uzs(dop_rasxod_total),
        }
    )

    return {
        "filters": filters,
        "summary": summary,
        "monthly": _build_monthly_series(metrics),
        "production_rows": production_rows,
        "updated_at": now_datetime().strftime("%H:%M"),
    }
